pkg/ingester/ingester.go

```diff
@@ -957,36 +982,56 @@ func (i *Ingester) pushSamplesToAppender(userID string, timeseries []mimirpb.Pre
 			switch cause := errors.Cause(err); cause {
 			case storage.ErrOutOfBounds:
 				stats.sampleOutOfBoundsCount++
-				updateFirstPartial(func() error { return newIngestErrSampleTimestampTooOld(model.Time(s.TimestampMs), ts.Labels) })
+				updateFirstPartial(func() error {
+					if ephemeral {
+						return newEphemeralIngestErrSampleTimestampTooOld(model.Time(s.TimestampMs), ts.Labels)
+					}
+					return newIngestErrSampleTimestampTooOld(model.Time(s.TimestampMs), ts.Labels)
+				})
 				continue
 
 			case storage.ErrOutOfOrderSample:
 				stats.sampleOutOfOrderCount++
-				updateFirstPartial(func() error { return newIngestErrSampleOutOfOrder(model.Time(s.TimestampMs), ts.Labels) })
+				updateFirstPartial(func() error {
+					if ephemeral {
+						return newEphemeralIngestErrSampleOutOfOrder(model.Time(s.TimestampMs), ts.Labels)
+					}
+					return newIngestErrSampleOutOfOrder(model.Time(s.TimestampMs), ts.Labels)
+				})
 				continue
 
 			case storage.ErrTooOldSample:
 				stats.sampleTooOldCount++
 				updateFirstPartial(func() error {
+					// OOO is not enabled for ephemeral storage, so we can't get this error.
 					return newIngestErrSampleTimestampTooOldOOOEnabled(model.Time(s.TimestampMs), ts.Labels, outOfOrderWindow)
 				})
 				continue
 
 			case storage.ErrDuplicateSampleForTimestamp:
 				stats.newValueForTimestampCount++
 				updateFirstPartial(func() error {
+					if ephemeral {
+						return newEphemeralIngestErrSampleDuplicateTimestamp(model.Time(s.TimestampMs), ts.Labels)
+					}
 					return newIngestErrSampleDuplicateTimestamp(model.Time(s.TimestampMs), ts.Labels)
 				})
 				continue
 
-			case errMaxSeriesPerUserLimitExceeded:
+			case errMaxSeriesPerUserLimitExceeded, errMaxEphemeralSeriesPerUserLimitExceeded: // we have special error for this, as we want different help message from FormatError.
 				stats.perUserSeriesLimitCount++
-				updateFirstPartial(func() error { return makeLimitError(perUserSeriesLimit, i.limiter.FormatError(userID, cause)) })
+				updateFirstPartial(func() error {
+					if ephemeral {
+						return makeLimitError(ephemeralDiscardPrefix+perUserSeriesLimit, i.limiter.FormatError(userID, cause))
```

Comment:
[nit] What if we define a const `ephemeralPerUserSeriesLimit` instead of `ephemeralDiscardPrefix+perUserSeriesLimit`? I've the feeling may be easier in the future to grep for it.